src/client/serialize/hdf_serialize.c

```diff
@@ -0,0 +1,1034 @@
+		goto out;
+	}
+	*cont_type = prop->dpp_entries[1].dpe_val;
+out:
+	return rc;
+}
+
+int
+deserialize_daos_cont_prop_metadata(char *filename, daos_prop_t *props,
+				    uint64_t *cont_type)
+{
+	int			rc = 0;
+	hid_t			file_id;
+
+	file_id = H5Fopen(filename, H5F_ACC_RDONLY, H5P_DEFAULT);
+	if (file_id < 0) {
+		fprintf(stderr, "failed to open hdf5 file");
+		rc = 1;
+		goto out;
+	}
+	rc = cont_deserialize_all_props(&file_id, props, cont_type);
+	if (rc != 0) {
+		fprintf(stderr, "failed to deserialize cont props: "
+			""DF_RC, DP_RC(rc));
+		goto out;
+	}
+out:
+	H5Fclose(file_id);
+	return rc;
+}
+
+static int
+cont_deserialize_usr_attrs(hid_t *file_id, uint64_t *_num_attrs,
+			   char ***_names, void ***_buffers, size_t **_sizes)
+{
+	hid_t		status = 0;
+	int		rc = 0;
+	int		num_attrs = 0;
+	int		num_dims = 0;
+	char		**names = NULL;
+	void		**buffers = NULL;
+	size_t		*sizes = NULL;
+	hid_t		dset = 0;
+	hid_t		dspace = 0;
+	hid_t		vtype = 0;
+	hsize_t		dims[1];
+	struct usr_attr	*attr_data = NULL;
+	int		i;
+
+	/* Read the user attributes */
+	dset = H5Dopen(*file_id, "User Attributes", H5P_DEFAULT);
+	if (dset < 0) {
+		fprintf(stderr, "failed to open user attributes dset");
+		rc = 1;
+		goto out;
+	}
+	dspace = H5Dget_space(dset);
+	if (dspace < 0) {
+		fprintf(stderr, "failed to get user attributes dspace");
+		rc = 1;
+		goto out;
+	}
+	vtype = H5Dget_type(dset);
+	if (vtype < 0) {
+		fprintf(stderr, "failed to get user attributes vtype");
+		rc = 1;
+		goto out;
+	}
+	num_dims = H5Sget_simple_extent_dims(dspace, dims, NULL);
+	if (num_dims < 0) {
+		fprintf(stderr, "failed to get user attributes dimensions");
+		rc = 1;
+		goto out;
+	}
+	num_attrs = dims[0];
+	attr_data = calloc(dims[0], sizeof(struct usr_attr));
+	if (attr_data == NULL) {
+		fprintf(stderr, "failed to allocate user attributes");
+		rc = 1;
+		goto out;
+	}
+	status = H5Dread(dset, vtype, H5S_ALL, H5S_ALL, H5P_DEFAULT, attr_data);
+	if (status < 0) {
+		fprintf(stderr, "failed to read user attributes data");
+		rc = 1;
+		goto out;
+	}
+	names = calloc(num_attrs, sizeof(char *));
+	if (names == NULL) {
+		fprintf(stderr, "failed to allocate user attributes");
+		rc = 1;
+		goto out;
+	}
+	buffers = calloc(num_attrs, sizeof(void *));
+	if (buffers == NULL) {
+		fprintf(stderr, "failed to allocate user attributes");
+		rc = 1;
+		goto out;
+	}
+	sizes = calloc(num_attrs, sizeof(size_t));
+	if (sizes == NULL) {
+		fprintf(stderr, "failed to allocate user attributes");
+		rc = 1;
+		goto out;
+	}
+	/* Set the user attribute buffers */
+	for (i = 0; i < num_attrs; i++) {
+		names[i] = attr_data[i].attr_name;
+		buffers[i] = attr_data[i].attr_val.p;
+		sizes[i] = attr_data[i].attr_val.len;
+	}
+	*_num_attrs = num_attrs;
+	*_names = names;
+	*_buffers = buffers;
+	*_sizes = sizes;
+out:
+	H5Dclose(dset);
+	H5Sclose(dspace);
+	H5Tclose(vtype);
+	free(attr_data);
+	return rc;
+}
+
+
+int
+deserialize_daos_cont_attrs_metadata(char *filename, uint64_t *num_attrs,
```

Comment:
We should probably have some naming convention for symbols exported by this library. Like daos_cont_serialize_md/attrs/...